worker/upload.go

```diff
@@ -204,102 +228,82 @@ func (w *worker) threadedUploadPackedSlabs(rs api.RedundancySettings, contractSe
-	// a slightly higher upload priority to avoid reaching the context deadline
-	if block {
-		_, err = w.uploadPackedSlabs(ctx, defaultPackedSlabsLockDuration, rs, contractSet, lockingPriorityBlockedUpload)
-	}
-
-	// make sure there's a goroutine uploading the remainder of the packed slabs
-	go w.threadedUploadPackedSlabs(rs, contractSet, lockingPriorityBackgroundUpload)
-	return
-}
-
-func (w *worker) uploadPackedSlabs(ctx context.Context, lockingDuration time.Duration, rs api.RedundancySettings, contractSet string, lockPriority int) (uploaded int, err error) {
 	// upload packed slabs
 	var mu sync.Mutex
 	var errs error
 
-	var wg sync.WaitGroup
-	totalSize := uint64(rs.TotalShards) * rhpv2.SectorSize
-
-	// derive a context that we can use as an interrupt in case of an error.
-	interruptCtx, cancel := context.WithCancel(ctx)
-	defer cancel()
+	// derive a context that we can use as an interrupt in case of an error or shutdown.
+	interruptCtx, interruptCancel := context.WithCancel(w.shutdownCtx)
+	defer interruptCancel()
 
+	var wg sync.WaitGroup
 	for {
-		// block until we have memory for a slab or until we are interrupted
-		mem := w.uploadManager.mm.AcquireMemory(interruptCtx, totalSize)
+		// block until we have memory
+		mem := w.uploadManager.mm.AcquireMemory(interruptCtx, rs.SlabSizeWithRedundancy())
 		if mem == nil {
 			break // interrupted
 		}
 
-		// fetch packed slabs to upload
-		var packedSlabs []api.PackedSlab
-		packedSlabs, err = w.bus.PackedSlabsForUpload(ctx, lockingDuration, uint8(rs.MinShards), uint8(rs.TotalShards), contractSet, 1)
-		if err != nil {
-			err = fmt.Errorf("couldn't fetch packed slabs from bus: %v", err)
-			mem.Release()
-			break
-		} else if len(packedSlabs) == 0 {
-			mem.Release()
-			break // no more slabs
-		}
-		ps := packedSlabs[0]
-
-		// launch upload for slab
 		wg.Add(1)
-		go func(ps api.PackedSlab) {
-			defer mem.Release()
+		go func() {
 			defer wg.Done()
-			err := w.uploadPackedSlab(ctx, rs, ps, mem, contractSet, lockPriority)
-			mu.Lock()
+			defer mem.Release()
+
+			// we use the background context here, but apply a sane timeout,
+			// this ensures ongoing uploads are handled gracefully during
+			// shutdown
+			ctx, cancel := context.WithTimeout(context.Background(), defaultPackedSlabsUploadTimeout)
+			defer cancel()
+
+			// attach interaction recorder to the context
+			ctx = context.WithValue(ctx, keyInteractionRecorder, w)
+
+			// try to upload a packed slab, if there were no packed slabs left to upload ok is false
+			ok, err := w.tryUploadPackedSlab(ctx, mem, defaultPackedSlabsLockDuration, rs, contractSet, lockPriority)
 			if err != nil {
+				mu.Lock()
 				errs = errors.Join(errs, err)
-				cancel() // prevent new uploads from being launched
-			} else {
-				uploaded++
+				mu.Unlock()
+				interruptCancel() // prevent new uploads from being launched
+			} else if !ok {
+				interruptCancel() // no more packed slabs to upload
```

Comment:
You can get rid of this `ok` by fetching the next slab for upload outside the goroutine, after `AcquireMemory`. If there are no more slabs, release the memory and break out of the loop.

The only reason we should ever need `interruptCancel` is to abort on error.